File: samples/language-impl-comparison/swift/basic_interpreter.swift
import Foundation

// MARK: - Value型
enum Value: Equatable {
    case number(Double)
    case string(String)
    case array([Value])
}

// MARK: - 演算子
enum BinOperator {
    case add, sub, mul, div
    case eq, ne, lt, le, gt, ge
    case and, or
}

enum UnaryOperator {
    case neg, not
}

// MARK: - 式
indirect enum Expr {
    case number(Double)
    case string(String)
    case variable(String)
    case arrayAccess(var: String, index: Expr)
    case binOp(op: BinOperator, left: Expr, right: Expr)
    case unaryOp(op: UnaryOperator, operand: Expr)
}

// MARK: - 文
indirect enum Statement {
    case let_(var: String, expr: Expr)
    case print([Expr])
    case if_(cond: Expr, thenBlock: [Statement], elseBlock: [Statement])
    case for_(var: String, start: Expr, end: Expr, step: Expr, body: [Statement])
    case while_(cond: Expr, body: [Statement])
    case goto(Int)
    case gosub(Int)
    case return_
    case dim(var: String, size: Expr)
    case end
}

// MARK: - プログラム型
typealias Program = [(line: Int, stmt: Statement)]

// MARK: - ランタイム状態
struct RuntimeState {
    var env: [String: Value]
    var callStack: [Int]
    var output: [String]
}

// MARK: - ランタイムエラー
enum RuntimeError: Error, CustomStringConvertible {
    case undefinedVariable(String)
    case undefinedLabel(Int)
    case typeMismatch(expected: String, got: String)
    case indexOutOfBounds
    case divisionByZero
    case stackUnderflow

    var description: String {
        switch self {
        case .undefinedVariable(let name):
            return "未定義変数: \(name)"
        case .undefinedLabel(let line):
            return "未定義ラベル: \(line)"
        case .typeMismatch(let expected, let got):
            return "型不一致: 期待 \(expected), 実際 \(got)"
        case .indexOutOfBounds:
            return "インデックス範囲外"
        case .divisionByZero:
            return "0で割ることはできません"
        case .stackUnderflow:
            return "スタックアンダーフロー"
        }
    }
}

// MARK: - インタープリタ実行
func run(program: Program) -> Result<[String], RuntimeError> {
    let state = RuntimeState(env: [:], callStack: [], output: [])
    let sorted = program.sorted { $0.line < $1.line }
    return executeProgram(sorted, pc: 0, state: state)
}

func executeProgram(_ program: Program, pc: Int, state: RuntimeState) -> Result<[String], RuntimeError> {
    guard pc < program.count else {
        return .success(state.output)
    }

    var newState = state
    let stmt = program[pc].stmt

    switch stmt {
    case .end:
        return .success(state.output)

    case .let_(let varName, let expr):
        switch evalExpr(expr, env: state.env) {
        case .success(let value):
            newState.env[varName] = value
            return executeProgram(program, pc: pc + 1, state: newState)
        case .failure(let error):
            return .failure(error)
        }

    case .print(let exprs):
        var parts: [String] = []
        for expr in exprs {
            switch evalExpr(expr, env: state.env) {
            case .success(let value):
                parts.append(valueToString(value))
            case .failure(let error):
                return .failure(error)
            }
        }
        newState.output.append(parts.joined(separator: " "))
        return executeProgram(program, pc: pc + 1, state: newState)

    case .if_(let cond, let thenBlock, let elseBlock):
        switch evalExpr(cond, env: state.env) {
        case .success(let condVal):
            let branch = isTruthy(condVal) ? thenBlock : elseBlock
            switch executeBlock(branch, state: state) {
            case .success(let blockState):
                return executeProgram(program, pc: pc + 1, state: blockState)
            case .failure(let error):
                return .failure(error)
            }
        case .failure(let error):
            return .failure(error)
        }

    case .for_(let varName, let start, let end, let step, let body):
        switch (evalExpr(start, env: state.env), evalExpr(end, env: state.env), evalExpr(step, env: state.env)) {
        case (.success(.number(let s)), .success(.number(let e)), .success(.number(let st))):
            return executeForLoop(var: varName, current: s, end: e, step: st, body: body, program: program, pc: pc, state: state)
        case (.failure(let error), _, _), (_, .failure(let error), _), (_, _, .failure(let error)):
            return .failure(error)
        default:
            return .failure(.typeMismatch(expected: "Number", got: "Other"))
        }

    case .while_(let cond, let body):
        return executeWhileLoop(cond: cond, body: body, program: program, pc: pc, state: state)

    case .goto(let target):
        switch findLine(program, target: target) {
        case .success(let newPc):
            return executeProgram(program, pc: newPc, state: state)
        case .failure(let error):
            return .failure(error)
        }

    case .gosub(let target):
        switch findLine(program, target: target) {
        case .success(let newPc):
            newState.callStack.append(pc + 1)
            return executeProgram(program, pc: newPc, state: newState)
        case .failure(let error):
            return .failure(error)
        }

    case .return_:
        guard let returnPc = state.callStack.last else {
            return .failure(.stackUnderflow)
        }
        newState.callStack.removeLast()
        return executeProgram(program, pc: returnPc, state: newState)

    case .dim(let varName, let sizeExpr):
        switch evalExpr(sizeExpr, env: state.env) {
        case .success(.number(let size)):
            let array = Array(repeating: Value.number(0.0), count: Int(size))
            newState.env[varName] = .array(array)
            return executeProgram(program, pc: pc + 1, state: newState)
        case .success(_):
            return .failure(.typeMismatch(expected: "Number", got: "Other"))
        case .failure(let error):
            return .failure(error)
        }
    }
}

func executeBlock(_ block: [Statement], state: RuntimeState) -> Result<RuntimeState, RuntimeError> {
    var currentState = state
    for stmt in block {
        switch executeSingleStatement(stmt, state: currentState) {
        case .success(let newState):
            currentState = newState
        case .failure(let error):
            return .failure(error)
        }
    }
    return .success(currentState)
}

func executeSingleStatement(_ stmt: Statement, state: RuntimeState) -> Result<RuntimeState, RuntimeError> {
    var newState = state
    switch stmt {
    case .let_(let varName, let expr):
        switch evalExpr(expr, env: state.env) {
        case .success(let value):
            newState.env[varName] = value
            return .success(newState)
        case .failure(let error):
            return .failure(error)
        }

    case .print(let exprs):
        var parts: [String] = []
        for expr in exprs {
            switch evalExpr(expr, env: state.env) {
            case .success(let value):
                parts.append(valueToString(value))
            case .failure(let error):
                return .failure(error)
            }
        }
        newState.output.append(parts.joined(separator: " "))
        return .success(newState)

    default:
        return .success(state)
    }
}

func executeForLoop(var varName: String, current: Double, end: Double, step: Double, body: [Statement], program: Program, pc: Int, state: RuntimeState) -> Result<[String], RuntimeError> {
    if (step > 0.0 && current > end) || (step < 0.0 && current < end) {
        return executeProgram(program, pc: pc + 1, state: state)
    }

    var newState = state
    newState.env[varName] = .number(current)

    switch executeBlock(body, state: newState) {
    case .success(let blockState):
        return executeForLoop(var: varName, current: current + step, end: end, step: step, body: body, program: program, pc: pc, state: blockState)
    case .failure(let error):
        return .failure(error)
    }
}

func executeWhileLoop(cond: Expr, body: [Statement], program: Program, pc: Int, state: RuntimeState) -> Result<[String], RuntimeError> {
    switch evalExpr(cond, env: state.env) {
    case .success(let condVal):
        if isTruthy(condVal) {
            switch executeBlock(body, state: state) {
            case .success(let newState):
                return executeWhileLoop(cond: cond, body: body, program: program, pc: pc, state: newState)
            case .failure(let error):
                return .failure(error)
            }
        } else {
            return executeProgram(program, pc: pc + 1, state: state)
        }
    case .failure(let error):
        return .failure(error)
    }
}

func evalExpr(_ expr: Expr, env: [String: Value]) -> Result<Value, RuntimeError> {
    switch expr {
    case .number(let n):
        return .success(.number(n))

    case .string(let s):
        return .success(.string(s))

    case .variable(let name):
        guard let value = env[name] else {
            return .failure(.undefinedVariable(name))
        }
        return .success(value)

    case .arrayAccess(let varName, let indexExpr):
        guard let arrayVal = env[varName] else {
            return .failure(.undefinedVariable(varName))
        }
        guard case .array(let arr) = arrayVal else {
            return .failure(.typeMismatch(expected: "Array", got: "Other"))
        }
        switch evalExpr(indexExpr, env: env) {
        case .success(.number(let idx)):
            let i = Int(idx)
            guard i >= 0 && i < arr.count else {
                return .failure(.indexOutOfBounds)
            }
            return .success(arr[i])
        case .success(_):
            return .failure(.typeMismatch(expected: "Number", got: "Other"))
        case .failure(let error):
            return .failure(error)
        }

    case .binOp(let op, let left, let right):
        switch (evalExpr(left, env: env), evalExpr(right, env: env)) {
        case (.success(let l), .success(let r)):
            return evalBinOp(op, left: l, right: r)
        case (.failure(let error), _), (_, .failure(let error)):
            return .failure(error)
        }

    case .unaryOp(let op, let operand):
        switch evalExpr(operand, env: env) {
        case .success(let val):
            return evalUnaryOp(op, operand: val)
        case .failure(let error):
            return .failure(error)
        }
    }
}

func evalBinOp(_ op: BinOperator, left: Value, right: Value) -> Result<Value, RuntimeError> {
    switch (op, left, right) {
    case (.add, .number(let l), .number(let r)):
        return .success(.number(l + r))
    case (.sub, .number(let l), .number(let r)):
        return .success(.number(l - r))
    case (.mul, .number(let l), .number(let r)):
        return .success(.number(l * r))
    case (.div, .number(let l), .number(let r)):
        guard r != 0.0 else { return .failure(.divisionByZero) }
        return .success(.number(l / r))
    case (.eq, .number(let l), .number(let r)):
        return .success(.number(l == r ? 1.0 : 0.0))
    case (.ne, .number(let l), .number(let r)):
        return .success(.number(l != r ? 1.0 : 0.0))
    case (.lt, .number(let l), .number(let r)):
        return .success(.number(l < r ? 1.0 : 0.0))
    case (.le, .number(let l), .number(let r)):
        return .success(.number(l <= r ? 1.0 : 0.0))
    case (.gt, .number(let l), .number(let r)):
        return .success(.number(l > r ? 1.0 : 0.0))
    case (.ge, .number(let l), .number(let r)):
        return .success(.number(l >= r ? 1.0 : 0.0))
    case (.and, let l, let r):
        return .success(.number(isTruthy(l) && isTruthy(r) ? 1.0 : 0.0))
    case (.or, let l, let r):
        return .success(.number(isTruthy(l) || isTruthy(r) ? 1.0 : 0.0))
    default:
        return .failure(.typeMismatch(expected: "Number", got: "Other"))
    }
}

func evalUnaryOp(_ op: UnaryOperator, operand: Value) -> Result<Value, RuntimeError> {
    switch (op, operand) {
    case (.neg, .number(let n)):
        return .success(.number(-n))
    case (.not, let v):
        return .success(.number(isTruthy(v) ? 0.0 : 1.0))
    default:
        return .failure(.typeMismatch(expected: "Number", got: "Other"))
    }
}

func isTruthy(_ value: Value) -> Bool {
    switch value {
    case .number(let n):
        return n != 0.0
    case .string(let s):
        return !s.isEmpty
    case .array(let a):
        return !a.isEmpty
    }
}

func valueToString(_ value: Value) -> String {
    switch value {
    case .number(let n):
        return String(n)
    case .string(let s):
        return s
    case .array(_):
        return "[Array]"
    }
}

func findLine(_ program: Program, target: Int) -> Result<Int, RuntimeError> {
    if let index = program.firstIndex(where: { $0.line == target }) {
        return .success(index)
    }
    return .failure(.undefinedLabel(target))
}

// MARK: - テスト実行例
let testProgram: Program = [
    (10, .let_(var: "x", expr: .number(0.0))),
    (20, .let_(var: "x", expr: .binOp(op: .add, left: .variable("x"), right: .number(1.0)))),
    (30, .print([.variable("x")])),
    (40, .if_(cond: .binOp(op: .lt, left: .variable("x"), right: .number(5.0)),
              thenBlock: [.goto(20)],
              elseBlock: [])),
    (50, .end)
]

switch run(program: testProgram) {
case .success(let output):
    output.forEach { print($0) }
case .failure(let error):
    print("エラー: \(error)")
}

File: samples/language-impl-comparison/rust/basic_interpreter.rs
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
enum Value {
    Number(f64),
    String(String),
    Array(Vec<Value>),
}

type Env = HashMap<String, Value>;

#[derive(Debug, Clone)]
enum Statement {
    Let { var: String, expr: Expr },
    Print(Vec<Expr>),
    If { cond: Expr, then_block: Vec<Statement>, else_block: Vec<Statement> },
    For { var: String, start: Expr, end: Expr, step: Expr, body: Vec<Statement> },
    While { cond: Expr, body: Vec<Statement> },
    Goto(i32),
    Gosub(i32),
    Return,
    Dim { var: String, size: Expr },
    End,
}

#[derive(Debug, Clone)]
enum Expr {
    Number(f64),
    String(String),
    Variable(String),
    ArrayAccess { var: String, index: Box<Expr> },
    BinOp { op: BinOperator, left: Box<Expr>, right: Box<Expr> },
    UnaryOp { op: UnaryOperator, operand: Box<Expr> },
}

#[derive(Debug, Clone, Copy)]
enum BinOperator {
    Add, Sub, Mul, Div,
    Eq, Ne, Lt, Le, Gt, Ge,
    And, Or,
}

#[derive(Debug, Clone, Copy)]
enum UnaryOperator {
    Neg, Not,
}

type Program = Vec<(i32, Statement)>;

struct RuntimeState {
    env: Env,
    call_stack: Vec<i32>,
    output: Vec<String>,
}

#[derive(Debug)]
enum RuntimeError {
    UndefinedVariable(String),
    UndefinedLabel(i32),
    TypeMismatch { expected: String, got: String },
    IndexOutOfBounds,
    DivisionByZero,
    StackUnderflow,
}

pub fn run(program: Program) -> Result<Vec<String>, RuntimeError> {
    let state = RuntimeState {
        env: HashMap::new(),
        call_stack: Vec::new(),
        output: Vec::new(),
    };

    let mut sorted = program;
    sorted.sort_by_key(|(line, _)| *line);

    execute_program(&sorted, 0, state)
}

fn execute_program(
    program: &[(i32, Statement)],
    pc: usize,
    mut state: RuntimeState,
) -> Result<Vec<String>, RuntimeError> {
    if pc >= program.len() {
        return Ok(state.output);
    }

    let (_line, stmt) = &program[pc];

    match stmt {
        Statement::End => Ok(state.output),

        Statement::Let { var, expr } => {
            let value = eval_expr(expr, &state.env)?;
            state.env.insert(var.clone(), value);
            execute_program(program, pc + 1, state)
        }

        Statement::Print(exprs) => {
            let values: Result<Vec<_>, _> = exprs.iter()
                .map(|e| eval_expr(e, &state.env))
                .collect();
            let text = values?
                .iter()
                .map(value_to_string)
                .collect::<Vec<_>>()
                .join(" ");
            state.output.push(text);
            execute_program(program, pc + 1, state)
        }

        Statement::If { cond, then_block, else_block } => {
            let cond_val = eval_expr(cond, &state.env)?;
            let branch = if is_truthy(&cond_val) { then_block } else { else_block };
            state = execute_block(branch, state)?;
            execute_program(program, pc + 1, state)
        }

        Statement::For { var, start, end, step, body } => {
            let start_val = eval_expr(start, &state.env)?;
            let end_val = eval_expr(end, &state.env)?;
            let step_val = eval_expr(step, &state.env)?;
            execute_for_loop(var, start_val, end_val, step_val, body, program, pc, state)
        }

        Statement::While { cond, body } => {
            execute_while_loop(cond, body, program, pc, state)
        }

        Statement::Goto(target) => {
            let new_pc = find_line(program, *target)?;
            execute_program(program, new_pc, state)
        }

        Statement::Gosub(target) => {
            let new_pc = find_line(program, *target)?;
            state.call_stack.push((pc + 1) as i32);
            execute_program(program, new_pc, state)
        }

        Statement::Return => {
            let return_pc = state.call_stack.pop()
                .ok_or(RuntimeError::StackUnderflow)?;
            execute_program(program, return_pc as usize, state)
        }

        Statement::Dim { var, size } => {
            let size_val = eval_expr(size, &state.env)?;
            match size_val {
                Value::Number(n) => {
                    let array = vec![Value::Number(0.0); n as usize];
                    state.env.insert(var.clone(), Value::Array(array));
                    execute_program(program, pc + 1, state)
                }
                _ => Err(RuntimeError::TypeMismatch {
                    expected: "Number".into(),
                    got: "Other".into(),
                }),
            }
        }
    }
}

fn execute_block(
    block: &[Statement],
    mut state: RuntimeState,
) -> Result<RuntimeState, RuntimeError> {
    for stmt in block {
        state = execute_single_statement(stmt, state)?;
    }
    Ok(state)
}

fn execute_single_statement(
    stmt: &Statement,
    mut state: RuntimeState,
) -> Result<RuntimeState, RuntimeError> {
    match stmt {
        Statement::Let { var, expr } => {
            let value = eval_expr(expr, &state.env)?;
            state.env.insert(var.clone(), value);
            Ok(state)
        }

        Statement::Print(exprs) => {
            let values: Result<Vec<_>, _> = exprs.iter()
                .map(|e| eval_expr(e, &state.env))
                .collect();
            let text = values?
                .iter()
                .map(value_to_string)
                .collect::<Vec<_>>()
                .join(" ");
            state.output.push(text);
            Ok(state)
        }

        _ => Ok(state),
    }
}

fn execute_for_loop(
    var: &str,
    start: Value,
    end: Value,
    step: Value,
    body: &[Statement],
    program: &[(i32, Statement)],
    pc: usize,
    state: RuntimeState,
) -> Result<Vec<String>, RuntimeError> {
    match (start, end, step) {
        (Value::Number(s), Value::Number(e), Value::Number(st)) => {
            for_loop_helper(var, s, e, st, body, program, pc, state)
        }
        _ => Err(RuntimeError::TypeMismatch {
            expected: "Number".into(),
            got: "Other".into(),
        }),
    }
}

fn for_loop_helper(
    var: &str,
    current: f64,
    end: f64,
    step: f64,
    body: &[Statement],
    program: &[(i32, Statement)],
    pc: usize,
    mut state: RuntimeState,
) -> Result<Vec<String>, RuntimeError> {
    if (step > 0.0 && current > end) || (step < 0.0 && current < end) {
        execute_program(program, pc + 1, state)
    } else {
        state.env.insert(var.to_string(), Value::Number(current));
        state = execute_block(body, state)?;
        for_loop_helper(var, current + step, end, step, body, program, pc, state)
    }
}

fn execute_while_loop(
    cond: &Expr,
    body: &[Statement],
    program: &[(i32, Statement)],
    pc: usize,
    state: RuntimeState,
) -> Result<Vec<String>, RuntimeError> {
    let cond_val = eval_expr(cond, &state.env)?;
    if is_truthy(&cond_val) {
        let new_state = execute_block(body, state)?;
        execute_while_loop(cond, body, program, pc, new_state)
    } else {
        execute_program(program, pc + 1, state)
    }
}

fn eval_expr(expr: &Expr, env: &Env) -> Result<Value, RuntimeError> {
    match expr {
        Expr::Number(n) => Ok(Value::Number(*n)),
        Expr::String(s) => Ok(Value::String(s.clone())),
        Expr::Variable(name) => {
            env.get(name)
                .cloned()
                .ok_or_else(|| RuntimeError::UndefinedVariable(name.clone()))
        }

        Expr::ArrayAccess { var, index } => {
            match env.get(var) {
                Some(Value::Array(arr)) => {
                    let idx_val = eval_expr(index, env)?;
                    match idx_val {
                        Value::Number(idx) => {
                            arr.get(idx as usize)
                                .cloned()
                                .ok_or(RuntimeError::IndexOutOfBounds)
                        }
                        _ => Err(RuntimeError::TypeMismatch {
                            expected: "Number".into(),
                            got: "Other".into(),
                        }),
                    }
                }
                Some(_) => Err(RuntimeError::TypeMismatch {
                    expected: "Array".into(),
                    got: "Other".into(),
                }),
                None => Err(RuntimeError::UndefinedVariable(var.clone())),
            }
        }

        Expr::BinOp { op, left, right } => {
            let l = eval_expr(left, env)?;
            let r = eval_expr(right, env)?;
            eval_binop(*op, l, r)
        }

        Expr::UnaryOp { op, operand } => {
            let val = eval_expr(operand, env)?;
            eval_unaryop(*op, val)
        }
    }
}

fn eval_binop(op: BinOperator, left: Value, right: Value) -> Result<Value, RuntimeError> {
    match (op, left, right) {
        (BinOperator::Add, Value::Number(l), Value::Number(r)) => Ok(Value::Number(l + r)),
        (BinOperator::Sub, Value::Number(l), Value::Number(r)) => Ok(Value::Number(l - r)),
        (BinOperator::Mul, Value::Number(l), Value::Number(r)) => Ok(Value::Number(l * r)),
        (BinOperator::Div, Value::Number(l), Value::Number(r)) => {
            if r == 0.0 {
                Err(RuntimeError::DivisionByZero)
            } else {
                Ok(Value::Number(l / r))
            }
        }
        (BinOperator::Eq, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l == r { 1.0 } else { 0.0 }))
        }
        (BinOperator::Ne, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l != r { 1.0 } else { 0.0 }))
        }
        (BinOperator::Lt, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l < r { 1.0 } else { 0.0 }))
        }
        (BinOperator::Le, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l <= r { 1.0 } else { 0.0 }))
        }
        (BinOperator::Gt, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l > r { 1.0 } else { 0.0 }))
        }
        (BinOperator::Ge, Value::Number(l), Value::Number(r)) => {
            Ok(Value::Number(if l >= r { 1.0 } else { 0.0 }))
        }
        (BinOperator::And, l, r) => {
            Ok(Value::Number(if is_truthy(&l) && is_truthy(&r) { 1.0 } else { 0.0 }))
        }
        (BinOperator::Or, l, r) => {
            Ok(Value::Number(if is_truthy(&l) || is_truthy(&r) { 1.0 } else { 0.0 }))
        }
        _ => Err(RuntimeError::TypeMismatch {
            expected: "Number".into(),
            got: "Other".into(),
        }),
    }
}

fn eval_unaryop(op: UnaryOperator, operand: Value) -> Result<Value, RuntimeError> {
    match (op, operand) {
        (UnaryOperator::Neg, Value::Number(n)) => Ok(Value::Number(-n)),
        (UnaryOperator::Not, v) => {
            Ok(Value::Number(if is_truthy(&v) { 0.0 } else { 1.0 }))
        }
        _ => Err(RuntimeError::TypeMismatch {
            expected: "Number".into(),
            got: "Other".into(),
        }),
    }
}

fn is_truthy(value: &Value) -> bool {
    match value {
        Value::Number(n) => *n != 0.0,
        Value::String(s) => !s.is_empty(),
        Value::Array(a) => !a.is_empty(),
    }
}

fn value_to_string(value: &Value) -> String {
    match value {
        Value::Number(n) => format!("{}", n),
        Value::String(s) => s.clone(),
        Value::Array(_) => "[Array]".to_string(),
    }
}

fn find_line(program: &[(i32, Statement)], target: i32) -> Result<usize, RuntimeError> {
    program
        .iter()
        .position(|(line, _)| *line == target)
        .ok_or(RuntimeError::UndefinedLabel(target))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_loop() {
        let program = vec![
            (10, Statement::Let { var: "x".into(), expr: Expr::Number(0.0) }),
            (20, Statement::Let {
                var: "x".into(),
                expr: Expr::BinOp {
                    op: BinOperator::Add,
                    left: Box::new(Expr::Variable("x".into())),
                    right: Box::new(Expr::Number(1.0)),
                },
            }),
            (30, Statement::Print(vec![Expr::Variable("x".into())])),
            (40, Statement::If {
                cond: Expr::BinOp {
                    op: BinOperator::Lt,
                    left: Box::new(Expr::Variable("x".into())),
                    right: Box::new(Expr::Number(3.0)),
                },
                then_block: vec![Statement::Goto(20)],
                else_block: vec![],
            }),
            (50, Statement::End),
        ];

        let result = run(program).unwrap();
        assert_eq!(result, vec!["1", "2", "3"]);
    }
}
